fix(ngcc): handle missing sources and out of range mappings when flattening source-maps

petebacondarwin opened the pull request:
This PR has two commits that fix issues with source-maps that are badly formatted.

---

If a package has a source-map but it does not provide the actual content of the sources, then the source-map flattening was crashing.

Now we ignore such mappings that have no source since we are not able to compute the merged mapping if there is no source file.

---

Previously when rendering flattened source-maps, it was assumed that no mapping would come from a line that is outside the lines of the actual source content. It turns out this is not a valid assumption.

Now the code that renders flattened source-maps will handle such mappings, with the additional benefit that the rendered source-map will only contain mapping lines up to the last mapping, rather than a mapping line for every content line.

---

Fixes #35709

gkalpak reviewed (approved):
LGTM :+1:

petebacondarwin commented:
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

petebacondarwin commented:
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

petebacondarwin commented:
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

JoostK commented on packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts:
```diff
@@ -290,11 +295,16 @@ export function parseMappings(
     for (const rawMapping of generatedLineMappings) {
       if (rawMapping.length >= 4) {
+        const originalSource = sources[rawMapping[1] !];
+        if (originalSource === null) {
```
Could it have been an out-of-bounds read in which case it would be `undefined`?

alxhub replied:
I don't think it could hurt here doing `==` instead of `===` to catch this case.

gkalpak commented:
BTW, it would be a good idea to run this against angular/ngcc-validation before merging to make sure there are no other regressions.

petebacondarwin committed:
refactor(ngcc): guard against a crash if source-map flattening fails

Source-maps in the wild could be badly formatted,
causing the source-map flattening processing to fail
unexpectedly. Rather than causing the whole of ngcc
to crash, we gracefully fallback to just returning the
generated source-map instead.

petebacondarwin committed:
fixup! fix(ngcc): handle missing sources when flattening source-maps

petebacondarwin committed:
fix(ngcc): handle missing sources when flattening source-maps

If a package has a source-map but it does not provide
the actual content of the sources, then the source-map
flattening was crashing.

Now we ignore such mappings that have no source
since we are not able to compute the merged
mapping if there is no source file.

Fixes #35709

petebacondarwin committed:
fix(ngcc): handle mappings outside the content when flattening source-maps

Previously when rendering flattened source-maps, it was assumed that no
mapping would come from a line that is outside the lines of the actual
source content. It turns out this is not a valid assumption.

Now the code that renders flattened source-maps will handle such
mappings, with the additional benefit that the rendered source-map
will only contain mapping lines up to the last mapping, rather than a
mapping line for every content line.

Fixes #35709

petebacondarwin commented:
Addressed the feedback. Now checking it against ngcc-validation repo.

JoostK commented on packages/compiler-cli/ngcc/src/rendering/source_maps.ts:
```diff
@@ -24,28 +25,39 @@ export interface SourceMapInfo {
+    }
+  } catch (e) {
+    logger.error(
+        `Failed to flatten the generated file source-map with the original source-maps failed: ${e.toString()}`);
```
```suggestion
        `Failed to flatten the generated file source-map with the original source-maps: ${e.toString()}`);
```

petebacondarwin replied:
Doh!
